roadmap/implementers-guide/src/node/disputes/votesdb.md

```diff
@@ -0,0 +1,137 @@
+# Votes DB
+
+Responsible module for tracking
+all kinds of votes on candidates
+by all validators for a fixed number
+of sessions.
+
+
+## Remarks
+
+Votes must be persisted. Disputes might happen long after
+there was a vote, as such just keeping things in memory
+is thus insufficient, as every blip (OOM, bug, maintanance)
+could cause the node to forget the state.
+
+## ToDo
+
+* Missing Definition of Governance Mode.
+* Is a `CommittedCandidateReceipt` enough or should the whole `PoV` be included? There was talk about detachable validator signatures.
+* Currently [`fn mark_bad()`](https://github.com/paritytech/substrate/pull/6301/files#diff-8faeb5c685a8fdff428c5ec6d9102fd59e127ff69762d43045cd38e586db5559R60-R64) does not persist data.
+
+## IO
+
+Inputs:
+
+* `VotesDbMessage::DisputeVote`
+
+Outputs:
+
+* `DisputeParticipationMessage::Detection`
+* `DisputeParticipationMessage::Resolution`
+
+## Messages
+
+```rust
+enum VotesDbMessage {
+    /// Allow querying all `Hash`es queried by a particular validator.
+    QueryValidatorVotes {
+        /// Validator indentification.
+        session: SessionIndex,
+        validator: ValidatorIndex,
+        response: ResponseChannel<Vec<Hash>>,
+    },
+
+    /// Store a vote for a particular dispute
+    StoreVote{
+        /// Unique validator indentification
+        session: SessionIndex,
+        validator: ValidatorIndex,
+        /// Vote.
+        vote: Vote,
+        /// Attestation.
+        attestation: ValidityAttestation,
+    },
+}
+```
+
+```rust
+enum DisputeMessage {
+    /// A dispute is detected
+    Detection {
+        /// unique validator indentification
+        session: SessionIndex,
+        validator: ValidatorIndex,
+        /// The attestation.
+        attestation: ValidityAttestation,
+        /// response channel
+        response: ResponseChannel<Vec<Hash>>,
+    },
+    /// Concluded a dispute with the following resolution
+    Resolution {
+        /// Resolution of the block can either be,
+        /// the block being valid or not.
+        valid: bool,
+    }
+}
+```
+
+## Helper Structs
+
+```rust
+/// Snapshot of the current vote state and recorded votes
+struct Snapshot {
+    /// all entries must be unique
+    pro: Vec<(ValidatorIndex, SessionIndex, ValidityAttestation)>,
+    /// all entries must be unique
+    cons: Vec<(ValidatorIndex, SessionIndex, ValidityAttestation)>,
+}
+```
+
+## Session Change
+
+A sweep clean is to be done to remove stored attestions
+of all validators that are not slashable / have no funds
```

Comment:
In fact, our dispute window is somewhat smaller than the window for slashing in general. We will just remove everything older than a fixed number of sessions e.g. a day's worth.l

Reply:
And that is due to the fact of the information spreads through the network (plus some generous margin) or is there another reason?

Reply:
It permits validators to clean up all the candidate blocks, sent messages, and erasure coded pieces they retain.

We need the validators to (a) support one another in checking candidates, but they're all done long before this, and (b) provide the XCMP messages sent by candidates they validated, meaning parachains must empty their sent messages within a day. 

In fact, we elsewhere argue that parachains should build a block per day or every 6 hour or something, but we never say they must empty they messages so fast, but actually this requires they do so, given we use sending checkers for message transport.

Reply:
It doesn't quite require that they act on the messages that fast, it just means that someone on the freceiving chain has to have got them, as they can't rely on validators to get them now.

Reply:
https://github.com/paritytech/polkadot/issues/2203